Sources/Utilities/SignInWithApple/SignInWithApple.swift

```diff
@@ -0,0 +1,200 @@
+    switch authorization.credential {
+    case let credential as ASAuthorizationAppleIDCredential:
+      guard let nonce = currentNonce else {
+        Logger.error("Missing nonce!")
+        delegate?.signInWithAppleDidFail(with: Error.missingNonce)
+        return
+      }
+      guard let identityToken = credential.identityToken else {
+        Logger.error("Missing identity token!")
+        delegate?.signInWithAppleDidFail(with: Error.missingIdentityToken)
+        return
+      }
+      guard let identityTokenString = String(data: identityToken, encoding: .utf8) else {
+        Logger.error("Invalid identity token!", params: ["error": identityToken.debugDescription])
+        delegate?.signInWithAppleDidFail(with: Error.invalidIdentityToken)
+        return
+      }
+      
+      // store userId for later
+      UserDefaults.standard.set(credential.user, forKey: userIdStorageKey)
+      
+      let displayName = [credential.fullName?.givenName, credential.fullName?.familyName]
+        .compactMap { $0 }
+        .joined(separator: " ")
+      
+      delegate?.signInWithAppleDidAuthorize(token: identityTokenString,
+                                                  nonce: nonce,
+                                                  displayName: displayName)
+    case _:
+      Logger.error("Unhandled authorization!")
+      delegate?.signInWithAppleDidFail(with: Error.unhandledAuthorization)
+    }
+  }
+  
+  func authorizationController(controller: ASAuthorizationController, didCompleteWithError error: Swift.Error) {
+    Logger.error("Sign in with Apple failed", params: ["error": error.localizedDescription])
+    delegate?.signInWithAppleDidFail(with: error)
+  }
+}
+
+// MARK: - ASAuthorizationControllerPresentationContextProviding
+@available(iOS 13.0, *)
+extension SignInWithApple: ASAuthorizationControllerPresentationContextProviding {
+  func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
+    presentationAnchor ?? UIWindow()
+  }
+}
+
+// MARK: - Private Methods
+@available(iOS 13.0, *)
+private extension SignInWithApple {
+  func generateRandomNonceString(length: Int = 32) -> String {
+    precondition(length > 0)
+    let charset: [Character] = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")
+    var result = ""
+    var remainingLength = length
+    
+    while remainingLength > 0 {
+      let randoms: [UInt8] = (0 ..< 16).map { _ in
+        var random: UInt8 = 0
+        let errorCode = SecRandomCopyBytes(kSecRandomDefault, 1, &random)
+        if errorCode != errSecSuccess {
+          fatalError("Unable to generate nonce. SecRandomCopyBytes failed with OSStatus \(errorCode)")
+        }
+        return random
+      }
+      
+      randoms.forEach {
+        guard length != 0 else { return }
+        if $0 < charset.count {
+          result.append(charset[Int($0)])
+          remainingLength -= 1
+        }
+      }
+    }
```

Comment:
Hm, do we need to do this, or can we get away by using randomElement()? So something like:
```swift
  let result = (0..<length).compactMap { _ in charset.randomElement() }
  guard result.count == length else {
    fatalError(...)
  }
  return String(result)
}
```
Not sure what the distributions are like in both cases, but the default (SystemRandomNumberGenerator) generator should be fine (according to docs):

> SystemRandomNumberGenerator is automatically seeded, is safe to use in multiple threads, and uses a cryptographically secure algorithm whenever possible.

Reply:
Never thought about this since the code was copy-pasted from Firebase [document](https://firebase.google.com/docs/auth/ios/apple). I'll take a look and see if we can make a replacement.

Reply:
Tested and it works. Thanks @nejcvivod 🙌